pinot-core/src/main/java/org/apache/pinot/core/query/reduce/DistinctDataTableReducer.java

```diff
@@ -65,13 +66,32 @@ public void reduceAndSetResults(String tableName, DataSchema dataSchema,
     // inside a DataTable
 
     // Gather all non-empty DistinctTables
+    // TODO: until we upgrade to newer version of pinot, we have to keep both code path. remove after 0.12.0 release.
+    // This is to work with server rolling upgrade when partially returned as DistinctTable Obj and partially regular
+    // DataTable; if all returns are DataTable we can directly merge with priority queue (with dedup).
     List<DistinctTable> nonEmptyDistinctTables = new ArrayList<>(dataTableMap.size());
     for (DataTable dataTable : dataTableMap.values()) {
-      DataTable.CustomObject customObject = dataTable.getCustomObject(0, 0);
-      assert customObject != null;
-      DistinctTable distinctTable = ObjectSerDeUtils.deserialize(customObject);
-      if (!distinctTable.isEmpty()) {
-        nonEmptyDistinctTables.add(distinctTable);
+      if (dataTable.getDataSchema().getColumnNames().length == 1
+          && dataTable.getDataSchema().getColumnDataType(0) == ColumnDataType.OBJECT) {
+        // distinctTable is still being returned as a single object
+        DataTable.CustomObject customObject = dataTable.getCustomObject(0, 0);
+        assert customObject != null;
+        DistinctTable distinctTable = ObjectSerDeUtils.deserialize(customObject);
+        if (!distinctTable.isEmpty()) {
+          nonEmptyDistinctTables.add(distinctTable);
+        }
+      } else {
+        // distinctTable is still being returned as normal data table.
+        if (dataTable.getNumberOfRows() > 0) {
+          List<Record> records = new ArrayList<>(dataTable.getNumberOfRows());
+          List<Object[]> rows =
+              SelectionOperatorUtils.reduceWithoutOrdering(Collections.singleton(dataTable), _queryContext.getLimit(),
```

Comment:
This is kind of confusing, and will create an extra list. Suggest adding a new method to directly read records from data table.

Reply:
👍